Implement a 'constants' arg

certik opened the pull request:
Fixes #11024.

certik committed:
Implement a 'constants' arg

asmeurer commented:
Can you add an example to the docstring (we should recommend using `numbered_symbols`). Other than that it looks good.

asmeurer commented on sympy/solvers/ode.py:
```diff
@@ -580,8 +583,17 @@ def dsolve(eq, func=None, hint="default", simplify=True,
+        # an actual iterator.
+        constants = iter(constants)
+        excluded_symbols = eq.atoms(Symbol)
+        constants = filter_symbols(constants, excluded_symbols)
```
I guess also document that it does this (if it isn't documented already).

certik commented:
Yes, I'll document it. Looks like there are still a few failures only triggered by the slow tests which I didn't run. I'll fix it.

certik committed:
Fix remaining constants arg

This fixes failures that were only triggered by "slow" tests, and thus I didn't
notice them in the previous commit.

certik committed:
Comment out failing tests

These assert statements currently fail.

certik commented:
@asmeurer so the problem is that previously the function `_helper_simplify` internally always generated the symbols starting from `C1`:
```
                    rv = _helper_simplify(eq, hint, hints[hint], simplify,
                            constants)
```
This function is called in a loop for each hint. Now the new version is using the `constants` iterator/generator, which just keeps increasing the symbol, so the first call to `_helper_simplify` draws a few symbols, say `C1, C2, C3`, just like in master. But the second call to `_helper_simplify` drfaws say `C1, C2` in master, but `C5, C6` in this PR.

This then causes the following code:
```
from sympy import *
var("x")
f = Function("f")
eq = x*f(x).diff(x) + f(x)
b = dsolve(eq, hint='all', simplify=False)
print(b['best'])
```
to print:
```
Eq(f(x), C7/x)
```
which causes some tests to fail, since they expect `Eq(f(x), C1/x)` instead.

Any ideas how to fix this? It looks like we need to wrap the generator into a custom class, that can be reset to 1 between each call to `_helper_simplify`. In other words, it would call the user defined generator to draw symbols `C1, C2, C3` and cache those, then when it gets reset, it will return the cached versions `C1, C2, C3`, and only when the next one is requested, it would go ahead and draw `C4` from the user provided generator.

asmeurer commented:
Yeah, I think that's the right solution (cache the output). Ideally you should only need n constants for an order n ODE, but I think there might be some solvers that use more than that (I could be wrong, though). If that were the case we could just `take(constants, n)` and use that everywhere.

vishalg2235 commented:
Currently in "test_ode" 8 tests are failed. out of which 7 were concerned about `ImportError: cannot import name dimensions` and one `AttributeError`. I think the conflicts must be resolved first or will it be fine if I import `dimensions` myself in branch ?

czgdp1807 commented:
@certik Please provide updates on this. The issue is still unresolved.
